### ISSUE TITLE
Wrong error reported when enum variant doesn't exist

### ISSUE THREAD

**Issue by StyMaar** [rust-lang/rust#51811]
Rustc doesn't report the right error when an enum variant doesn't exist and the enum variant contains an anonymous struct. It works well if the enum variant contains a value, or if it's an empty enum variant.

```Rust

enum Test {
    One,
    #[cfg(test)]
    Two{
        foo: String,
    },
    #[cfg(test)]
    Three,
     #[cfg(test)]
     Four(String)
}  

fn main() {
    works();
    works_too();
    fails();
}


fn works(){
    let y = Test::Three; // here we have the expected error: `variant not found in `Test``
}

fn works_too(){
    let y = Test::Four("yolo".to_owned()); // here we have the expected error: `variant not found in `Test``
}

fn fails(){
    let y = Test::Two{foo: "yolo".to_owned()}; // here, the error is not what it should be (`ambiguous associated type`)
}
```

## Meta

`rustc --version --verbose`:

```
rustc 1.28.0-nightly (2a1c4eec4 2018-06-25)
binary: rustc
commit-hash: 2a1c4eec40527de45b9d9b81672c8b9220d554fc
commit-date: 2018-06-25
host: x86_64-unknown-linux-gnu
release: 1.28.0-nightly
LLVM version: 6.0
```

But it can also be reproduced on the current stable (1.27) on the playground : https://play.rust-lang.org/?gist=d998c11b1ebb7825b328930db144eb8e&version=stable&mode=debug

**Comment by estebank** [comment]
Current output:

```
error[E0599]: no variant named `Three` found for type `Test` in the current scope
  --> src/main.rs:21:13
   |
1  | enum Test {
   | --------- variant `Three` not found here
...
21 |     let y = Test::Three; // here we have the expected error: `variant not found in `Test``
   |             ^^^^^^^^^^^ variant not found in `Test`

error[E0599]: no variant named `Four` found for type `Test` in the current scope
  --> src/main.rs:25:13
   |
1  | enum Test {
   | --------- variant `Four` not found here
...
25 |     let y = Test::Four("yolo".to_owned()); // here we have the expected error: `variant not found in `Test``
   |             ^^^^^^^^^^ variant not found in `Test`

error: no variant `Two` on enum `Test`
  --> src/main.rs:29:13
   |
29 |     let y = Test::Two{foo: "yolo".to_owned()}; // here, the error is not what it should be (`ambiguous associated type`)
   |             ^^^^^^^^^ unknown variant

error: aborting due to 3 previous errors
```

**Comment by StyMaar** [comment]
Cool. I have one question though: Why is the error message different ? “unknown variant” vs “variant not found in `Test`”?

**Comment by estebank** [comment]
@StyMaar because it is being checked in two different parts of the code. FWIW, the output is now slightly more consistent:

```
error[E0599]: no variant or associated item named `Three` found for type `Test` in the current scope
  --> src/main.rs:21:19
   |
1  | enum Test {
   | --------- variant or associated item `Three` not found here
...
21 |     let y = Test::Three; // here we have the expected error: `variant not found in `Test``
   |                   ^^^^^ variant or associated item not found in `Test`

error[E0599]: no variant or associated item named `Four` found for type `Test` in the current scope
  --> src/main.rs:25:19
   |
1  | enum Test {
   | --------- variant or associated item `Four` not found here
...
25 |     let y = Test::Four("yolo".to_owned()); // here we have the expected error: `variant not found in `Test``
   |                   ^^^^ variant or associated item not found in `Test`

error: no variant `Two` in enum `Test`
  --> src/main.rs:29:19
   |
1  | enum Test {
   | --------- variant `Two` not found here
...
29 |     let y = Test::Two{foo: "yolo".to_owned()}; // here, the error is not what it should be (`ambiguous associated type`)
   |             ------^^^
   |             |
   |             variant not found in `Test`

error: aborting due to 3 previous errors
```
